runtime/runtime_linux.go

```diff
@@ -2,39 +2,220 @@ package runtime
+	"io/ioutil"
 	"log"
+	"regexp"
+	"strings"
 	"time"
+
+	"github.com/taskcluster/generic-worker/host"
 )
 
 func (user *OSUser) CreateNew(okIfExists bool) (err error) {
-	return errors.New("(*(runtime.OSUser)).CreateNew(bool) not implemented on Linux")
+	if okIfExists {
+		panic("(*(runtime.OSUser)).CreateNew(true) not implemented on linux")
+	}
+
+	createUserScript := `
+		set -eu
+		username="${0}"
+		homedir="/home/${0}"
+		password="${1}"
+		echo "Creating user '${username}' with home directory '${homedir}' and password '${password}'..."
+		/usr/bin/sudo /usr/sbin/adduser --disabled-password --gecos "" --debug --home "${homedir}" "${username}"
+		echo "${username}:${password}" | /usr/bin/sudo /usr/sbin/chpasswd
+	`
+
+	return host.Run("/bin/bash", "-c", createUserScript, user.Name, user.Password)
 }
 
 func DeleteUser(username string) (err error) {
-	return errors.New("runtime.DeleteUser(string) not implemented on Linux")
+	return host.Run("/usr/bin/sudo", "/usr/sbin/deluser", "--force", "--remove-all-files", username)
 }
 
 func ListUserAccounts() (usernames []string, err error) {
-	return nil, errors.New("runtime.ListUserAccounts() not implemented on Linux")
+	var passwd []byte
+	passwd, err = ioutil.ReadFile("/etc/passwd")
+	if err != nil {
+		return
+	}
+	lines := strings.Split(string(passwd), "\n")
+	usernames = []string{}
+	for _, line := range lines {
+		if line != "" {
+			usernames = append(usernames, strings.SplitN(line, ":", 2)[0])
+		}
+	}
+	return
 }
 
 func UserHomeDirectoriesParent() string {
 	return "/home"
 }
 
 func WaitForLoginCompletion(timeout time.Duration) error {
-	return errors.New("runtime.WaitForLoginCompletion(time.Duration) not implemented on Linux")
+	deadline := time.Now().Add(timeout)
+	log.Print("Checking if user is logged in...")
+	for time.Now().Before(deadline) {
+		_, err := InteractiveUsername()
+		if err != nil {
+			log.Printf("WARNING: Error checking for interactive user: %v", err)
+			time.Sleep(time.Second)
+			continue
+		}
+		return nil
+	}
+	log.Print("Timed out waiting for user login")
+	return errors.New("No user logged in with console session")
 }
 
 func InteractiveUsername() (string, error) {
-	return "", errors.New("runtime.InteractiveUsername() not implemented on Linux")
+	gnomeSessionUserList, err := host.CombinedOutput("/bin/bash", "-c", "PROCPS_USERLEN=20 /usr/bin/w -s | /bin/grep gnome-[s]ession | /usr/bin/cut -f1 -d' '")
+	if err != nil {
+		return "", fmt.Errorf("Cannot run command to determine the interactive user: %v", err)
+	}
+	lines := strings.Split(gnomeSessionUserList, "\n")
+	if len(lines) != 2 || lines[1] != "" {
+		return "", fmt.Errorf("Number of gnome session users is not exactly one - not sure which user is interactively logged on: %#v", lines)
+	}
+	return lines[0], nil
+}
+
+type sectionHandler interface {
+	lineInsideSection(line string)
+	lineOutsideSection(line string)
+}
+
+func iniFileLineHandler(iniFile string, section string, sh sectionHandler) (err error) {
```

Comment:
Does this kind of logic make sense to handle in a subpackage?

Reply:
That's a good idea. I might slightly adapt/simplify it too at the same time.